flax/experimental/nnx/nnx/nn/attention.py

```diff
@@ -0,0 +1,669 @@
+          "MultiHeadAttention's `__call__` method "
+          'has changed to `__call__(inputs_q, inputs_k=None, '
+          'inputs_v=None, *, inputs_kv=None, mask=None, '
+          'deterministic=None)`. Use the kwarg `mask` instead. '
+          'See https://github.com/google/flax/discussions/3389 '
+          'and read the docstring for more information.',
+          DeprecationWarning,
+        )
+
+    if inputs_q.shape[-1] != self.features_in:
+      raise ValueError(
+        f'Incompatible input dimension, got {inputs_q.shape[-1]} '
+        f'but module expects {self.features_in}.'
+      )
+
+    query = self.query(inputs_q)
+    key = self.key(inputs_k)
+    value = self.value(inputs_v)
+
+    if self.normalize_qk:
+      assert self.query_ln is not None and self.key_ln is not None
+      # Normalizing query and key projections stabilizes training with higher
+      # LR. See ViT-22B paper http://arxiv.org/abs/2302.05442 for analysis.
+      query = self.query_ln(query)
+      key = self.key_ln(key)
+
+    # During fast autoregressive decoding, we feed one position at a time,
+    # and cache the keys and values step by step.
+    if self.decode:
+      (
+        *batch_dims,
+        max_length,
+        num_heads,
+        depth_per_head,
+      ) = self.cached_key.shape
+      # shape check of cached keys against query input
+      expected_shape = tuple(batch_dims) + (1, num_heads, depth_per_head)
+      if expected_shape != query.shape:
+        raise ValueError(
+          'Autoregressive cache shape error, '
+          'expected query shape %s instead got %s.'
+          % (expected_shape, query.shape)
+        )
+      # update key, value caches with our new 1d spatial slices
+      cur_index = self.cache_index
+      indices = (0,) * len(batch_dims) + (cur_index, 0, 0)
+      key = lax.dynamic_update_slice(self.cached_key, key, indices)
+      value = lax.dynamic_update_slice(self.cached_value, value, indices)
+      self.cached_key = key
+      self.cached_value = value
+      self.cache_index += 1
+      # causal mask for cached decoder self-attention:
+      # our single query position should only attend to those key
+      # positions that have already been generated and cached,
+      # not the remaining zero elements.
+      mask = combine_masks(
+        mask,
+        jnp.broadcast_to(
+          jnp.arange(max_length) <= cur_index,
+          tuple(batch_dims) + (1, 1, max_length),
+        ),
+      )
+
+    if (
+      self.dropout_rate > 0.0
+    ):  # Require `deterministic` only if using dropout.
+      m_deterministic = first_from(
+        'deterministic',
+        deterministic,
+        self.deterministic,
+        flags.get('deterministic'),
+      )
+      if not m_deterministic and dropout_rng is None:
+        if rngs is None:
+          raise ValueError(
+            "'rngs' must be provided if 'dropout_rng' is not given."
+          )
+        dropout_rng = rngs.dropout()
```

Comment:
In Flax, `dropout_rng` is passed in by the user and used to generate the dropout mask. Here, it looks like `dropout_rng` is being overwritten. Is `rngs.dropout` similar/equivalent to `self.make_rng`? If so, maybe `dropout_rng` isn't needed as an arg and we just pass in an `rngs` arg?